### PR TITLE
feat: Introduce MFA recipe in plugin interface

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Introduce MFA recipe

## Related issues
- https://github.com/supertokens/supertokens-core/issues/648

## Checklist for important updates
- [x] Changelog has been updated
- [ ] Changes to the version if needed
   - In `build.gradle`
- [x] Had installed and ran the pre-commit hook
- [ ] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

### CODE REVIEW

**Review by rishabhpoddar — APPROVED** [review]

**Comment by KShivendu** [comment]
Please check the latest change before merging. https://github.com/supertokens/supertokens-plugin-interface/pull/87/commits/77a78b0bc6cb227afb5bd315f73bfb51a38c22f7

File: src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    boolean enableFactor(TenantIdentifier tenantIdentifier, String userId, String factorId) throws StorageQueryException;
+
+    // List all the factors for a user:
+    String[] listFactors(TenantIdentifier tenantIdentifier, String userId) throws StorageQueryException;
+
+    // Disable (delete) a factor for a user and return true if it actually deleted something.
+    boolean disableFactor(TenantIdentifier tenantIdentifier, String userId, String factorId) throws StorageQueryException;
+
+    // Delete a user and all of their factors.
+    boolean deleteUser(AppIdentifier tenantIdentifier, String userId) throws StorageQueryException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java:19]: you will also need a function to delete mfa user's for a specific tenant

Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java:19]:
```suggestion
    boolean deleteUser(AppIdentifier appIdentifier, String userId) throws StorageQueryException;
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java (hunk 1 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java:19]: fixed.

Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java]: Added

File: src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java [hunk 2]
```diff
@@ -15,6 +15,9 @@ public interface MfaStorage extends NonAuthRecipeStorage {
     // Disable (delete) a factor for a user and return true if it actually deleted something.
     boolean disableFactor(TenantIdentifier tenantIdentifier, String userId, String factorId) throws StorageQueryException;
 
-    // Delete a user and all of their factors.
-    boolean deleteUser(AppIdentifier tenantIdentifier, String userId) throws StorageQueryException;
+    // Delete a user across all tenants (with all the relevant factors)
+    boolean deleteUser(AppIdentifier appIdentifier, String userId) throws StorageQueryException;
+
+    // Delete a user from a tenant (with all the relevant factors)
+    boolean deleteUserFromTenant(TenantIdentifier tenantIdentifier, String userId) throws StorageQueryException;
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java:22]:
```suggestion
    boolean deleteMFAUser(AppIdentifier appIdentifier, String userId) throws StorageQueryException;

    // Delete a user from a tenant (with all the relevant factors)
    boolean deleteMFAUser(TenantIdentifier tenantIdentifier, String userId) throws StorageQueryException;
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java (hunk 2 above)
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/mfa/MfaStorage.java:22]: Done

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

File: src/main/java/io/supertokens/pluginInterface/ActiveUsersStorage.java [hunk 3]
```diff
@@ -17,5 +17,12 @@ public interface ActiveUsersStorage extends NonAuthRecipeStorage {
     /* Count the number of users who have enabled TOTP and are active */
     int countUsersEnabledTotpAndActiveSince(AppIdentifier appIdentifier, long time) throws StorageQueryException;
 
+    /* Count the number of users who have enabled MFA */
```
Comment (KShivendu) [inline src/main/java/io/supertokens/pluginInterface/ActiveUsersStorage.java:20]: What's the use of `jar/plugin-interface-2.23.0.jar`?  If none, remove it.

[the remark this answers is not in the document]
Comment (KShivendu) [reply src/main/java/io/supertokens/pluginInterface/ActiveUsersStorage.java:20]: Compare the current PR with https://github.com/supertokens/supertokens-plugin-interface/pull/87/files/a67d94df05fb816a4695a1e5309daded3c26d570..6e5f1ec5d2281af9e596f48e875f34540716044b once

**Pull request merged**